File: array/MatrixMultiplication/python/MatrixMultiplication.py
from __future__ import annotations

Matrix = list[list[int]]


def multiply_matrices(left: Matrix, right: Matrix) -> Matrix:
    if not left or not right:
        raise ValueError("matrices must be non-empty")

    rows = len(left)
    shared = len(left[0])
    right_rows = len(right)
    cols = len(right[0])

    if shared == 0 or cols == 0 or shared != right_rows:
        raise ValueError("matrix dimensions are incompatible")
    if any(len(row) != shared for row in left):
        raise ValueError("left matrix rows have inconsistent lengths")
    if any(len(row) != cols for row in right):
        raise ValueError("right matrix rows have inconsistent lengths")

    result = [[0 for _ in range(cols)] for _ in range(rows)]
    for i in range(rows):
        for k in range(shared):
            value = left[i][k]
            for j in range(cols):
                result[i][j] += value * right[k][j]
    return result

File: array/MatrixMultiplication/rust/src/matrix_multiplication.rs
pub type Matrix = Vec<Vec<i64>>;

pub fn multiply_matrices(left: &Matrix, right: &Matrix) -> Result<Matrix, String> {
    if left.is_empty() || right.is_empty() {
        return Err("matrices must be non-empty".to_string());
    }

    let rows = left.len();
    let shared = left[0].len();
    let right_rows = right.len();
    let cols = right[0].len();

    if shared == 0 || cols == 0 || shared != right_rows {
        return Err("matrix dimensions are incompatible".to_string());
    }
    if left.iter().any(|row| row.len() != shared) {
        return Err("left matrix rows have inconsistent lengths".to_string());
    }
    if right.iter().any(|row| row.len() != cols) {
        return Err("right matrix rows have inconsistent lengths".to_string());
    }

    let mut result = vec![vec![0_i64; cols]; rows];
    for i in 0..rows {
        for k in 0..shared {
            let value = left[i][k];
            for j in 0..cols {
                result[i][j] += value * right[k][j];
            }
        }
    }
    Ok(result)
}
